design/external-references.md

```diff
@@ -0,0 +1,344 @@
+
+A common example is for a common infrastructure team to maintain a CDK app which defines networking resources such as VPCs, security groups, LBs, etc and application teams need to reference these resources in their own CDK apps.
+
+__So the problem we are trying to solve is how to enable users to reference resources (and potentially higher level constructs) across CDK apps.__
+
+## Scope
+
+As an initial step, we'll address the use case of referencing resources across apps but within __the same AWS environment__ (account + region). This limitation is based on the limitation AWS CloudFormation has for exports/imports.
+
+However, there are definitely use cases for being able to reference resources across environments, and we are also aware that some customers are reluctant to use CloudFormation import/exports due to the strong-coupling this mechanism creates between the stacks (for example, if an Output of one stack is being referenced by another stack, the original stack cannot be deleted).
+
+Another related use case is referencing resources by runtime code. An application's runtime code (AWS Lambda functions, CodeBuild projects, ECS containers, etc) needs to be able to interact with resources defined in CDK apps (e.g. read/write from a Amazon DynamoDB table, send messages to an SQS Queue, etc). This is a very similar use case to the cross-app references because basically the same information needs to be transferred between the apps.
+
+## Requirements
+
+- **REQ1**: Allow constructs to be published from one CDK app and consumed by other CDK apps within the same environment
+- **REQ2**: Support "simple" constructs, which are resources that can be represented by a set of attributes such as an ARN
+- Support "composite" constructs, which are resources that encapsulate other resources (such as security groups, roles, etc).
+- **REQ3**: Allow CDK apps to reference resources that were created elsewhere (by other CDK apps, manually or any other mean) by referencing the physical identity. For example, an S3 Bucket ARN should be sufficient in order to reference
+an existing S3 bucket.
+- **REQ4**: If possible, implement the mechanism such that it can be later used to publish and consume constructs through other key/value mechanism such as environment variables, SSM parameters, etc.
+- **REQ5**: If a resource is exported twice under the same export name, only a single set of outputs will be created ([#1496](https://github.com/awslabs/aws-cdk/issues/1496)).
+- **REQ6**: If the same resource is imported twice in the same stack, the "import" method will return the same object.
+- **REQ7**: For some resources (like VPC for example), it should be possible to import the resource by retrieving it's details from the deployed environment via a context provider.
+
```

Comment:
REQ8: It must be possible for some constructs to retrieve the import values at *synthesis* time, so that CDK can execute some code for each element in a variable length list (in particular, for objects that have a set of Security Groups).

Reply:
Yes, I am actually addressing later but I’ll add here. You mean through context providers right?

Reply:
I don't mean anything yet. I was hoping this mechanism might also work at synthesis time. But if you're explicitly saying it doesn't and never will, then yes, our only other option is to write a context provider.

But even then, I think I'd want a generic context provider mechanism, because otherwise we'll have to write a toolkit plugin for every construct library that happens to use EC2 compute (which I imagine is going to be a bunch of them).

Reply:
I know I'm coming across belligerent, but I really see this as a very important aspect of the mechanism, and I feel we cannot conclude this discussion until we have a solid answer for transporting synthesis-time information.

Maybe the answer is that we have 2 classes of exports, one which resolves deployment time and one which resolves synthesis-time. It might be transparent to the coder, but visible to the synthesis mechanism.

Reply:
The output mechanism doesn't have to change for this, only the time of reading needs to be configurable.

Reply:
I see what you are saying and I agree. Let’s think of a way to make synthesis time imports work seamlessly using this mechanism. Let me think about it. I think you have the right intuition that this can work.

Reply:
Only thing we'd need is a context provider for stack/output/export values I think.

Although we'd have to contend with the fact that these values will be cached, and now users would have to be aware of the fact that they have to invalidate them.